server/service/hosts.go

```diff
@@ -136,67 +136,32 @@ func listHostsEndpoint(ctx context.Context, request interface{}, svc fleet.Servi
 	}, nil
 }
 
-func (svc *Service) AggregatedMDMSolutions(ctx context.Context, teamID *uint, mdmID uint) (*fleet.AggregatedMDMSolutions, error) {
-	if err := svc.authz.Authorize(ctx, &fleet.Host{TeamID: teamID}, fleet.ActionList); err != nil {
-		return nil, err
-	}
-
-	if teamID != nil {
-		_, err := svc.ds.Team(ctx, *teamID)
-		if err != nil {
-			return nil, err
-		}
-	}
-
-	// it is expected that there will be relatively few MDM solutions. This
-	// returns the slice of all aggregated stats (one entry per mdm_id), and we
-	// then iterate to return only the one that was requested (the slice is
-	// stored as-is in a JSON field in the database).
-	sols, _, err := svc.ds.AggregatedMDMSolutions(ctx, teamID)
-	if err != nil {
+func (svc *Service) GetMDMSolution(ctx context.Context, mdmID uint) (*fleet.MDMSolution, error) {
+	// require list hosts permission to view this information
+	if err := svc.authz.Authorize(ctx, &fleet.Host{}, fleet.ActionList); err != nil {
 		return nil, err
 	}
-
-	for _, sol := range sols {
-		// don't take the address of the loop variable (although it could be ok
-		// here, but just bad practice)
-		sol := sol
-		if sol.ID == mdmID {
-			return &sol, nil
-		}
+	sol, err := svc.ds.GetMDMSolution(ctx, mdmID)
+	if err != nil && fleet.IsNotFound(err) {
+		// ignore not found, just return nil (we don't want the caller - e.g. List
+		// Hosts - to return 404 because of that)
+		return nil, nil
```

Comment:
this (and same below) sounds like a responsibility for the caller, but I don't have strong feelings about it.

Reply:
Great point, I'll move that to the endpoint handler function!